dev/prepare-merge-commands.sh

```diff
@@ -0,0 +1,207 @@
+echo "git fetch $repository"
+
+# Get a subject from the first commit which will serve as a title to be pasted into CHANGES.txt
+first_commit="$(git log --pretty=format:%s --reverse "$apache_repository/${target_branches[0]}..$repository/${feature_branches[0]}" | head -n 1)"
+
+push_command="git push --atomic $apache_repository"
+skipped_branches_found=0
+for i in $(seq 0 $((${#target_branches[@]} - 1))); do
+  echo ""
+  echo ""
+  echo ""
+  echo "# $repository/${feature_branches[$i]} -> ${target_branches[$i]}"
+  echo "# --------------------------------------------------------------------------------------------------------"
+
+  if [[ $i == 0 ]] && [[ "${feature_branches[$i]}" == "" ]]; then
+    # Although we can skip a feature for some versions, we cannot skip it for the oldest version (which is quite obvious)
+    exit 1
+  fi
+
+  # Read the list of commits between the remote head and the feature branch - we need to cherry pick them (or some of them)
+
+  echo "git switch ${target_branches[$i]}"
+  echo "git reset --hard $apache_repository/${target_branches[$i]}"
+  if [[ "${feature_branches[$i]}" == "" ]]; then
+    skipped_branches_found=1
+    # A script for the case where there is no fix for a version
+    echo "git merge -s ours --log --no-edit ${target_branches[$((i - 1))]}"
+  else
+    readarray -t commits < <(git log --reverse --oneline "$apache_repository/${target_branches[$i]}..$repository/${feature_branches[$i]}")
+
+    if [[ $i != 0 ]]; then
+      # When this isn't the oldest version (we want to have only the merge commit)
+      echo "git merge -s ours --log --no-edit ${target_branches[$((i - 1))]}"
+    fi
+
+    for c in $(seq 0 $((${#commits[@]} - 1))); do
+      commit_sha="$(echo "${commits[$c]}" | cut -f 1 -d ' ')"
+      if [[ $i == 0 ]] && [[ $c == 0 ]]; then
+        # we want to have only one feature commit (c == 0), which is in the oldest version (i == 0)
+        echo "git cherry-pick $commit_sha # ${commits[$c]}"
+      else
+        # otherwise we squash the commits to the previous one
+        echo "git cherry-pick -n $commit_sha && git commit -a --amend --no-edit # ${commits[$c]}"
+      fi
+    done
+  fi
+
+  if [[ "$skipped_branches_found" == "0" ]]; then
+    if [[ $i == 0 ]]; then
+      echo "grep '$ticket' CHANGES.txt || sed -E -i '/^[0-9]+\.[0-9]+/{s/.*/&\n\ * $first_commit ($ticket)/;:a;n;ba}' CHANGES.txt"
+    else
+      echo "grep '$ticket' CHANGES.txt || sed -E -i '/^Merged from ${oldest_target_version}/{s/.*/&\n\ * $first_commit ($ticket)/;:a;n;ba}' CHANGES.txt"
+    fi
+    echo "git diff CHANGES.txt"
+  else
+    echo "Update CHANGES.txt by adding the following line:"
+    echo " * $first_commit ($ticket)"
+  fi
+  echo "git add CHANGES.txt"
+  echo "git commit --amend --no-edit"
+
+  echo "(git diff $apache_repository/${target_branches[$i]}..HEAD -- .circleci/ | git apply -R --index) && git commit -a --amend --no-edit # Remove all changes in .circleci directory if you need to"
```

Comment:
nit: i'm a bit uneasy about this. i think it's simpler and safer if we just expect the user to remove the throwaway circleci commits before calling this script. (though we should add a safety catch that fails if such a commit exists.)